Add pre-publish preview links

alecgeatches opened the pull request:
## Description

Add "Secure Preview" option to the sidebar of private posts:

![secure-link-demo](https://github.com/user-attachments/assets/3321ce89-5c47-4931-b6af-54b80f4a7636)

## Implementation

### Token generation

General token logic was borrowed from the [VIP Decoupled Bundle's `Token` class](https://github.com/Automattic/vip-decoupled-bundle/blob/trunk/preview/token.php). The implementation in this plugin has been changed in a handful of ways:

- Instead of specifying global values for [expiration time](https://github.com/Automattic/vip-decoupled-bundle/blob/7c26592f4651777f13b87b51fef3319f99a9ce4e/preview/token.php#L26) and [whether tokens expire on use](https://github.com/Automattic/vip-decoupled-bundle/blob/7c26592f4651777f13b87b51fef3319f99a9ce4e/preview/token.php#L136), these parameters [are now options](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/token.php#L26) provided to token generation. There's still an existing check that a token lifetime doesn't exceed a maximum amount, but this can be [modified via the `vw_secure_preview_max_token_lifetime` filter](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/token.php#L113) similar to the decoupled implementation.
- Link lifetimes are configured in [`get_link_expiration_options()`](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/secure-preview.php#L121-L149) and can be modified via filter to add shorter or longer preview times. This can be combined with the `vw_secure_preview_max_token_lifetime` filter above to set custom expiration times.
- The logic for one-time-use tokens was determined by filter in the decoupled plugin, but now is determined by individual token. [`validate_token` logic has changed](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/token.php#L73) to handle this stored parameter.

While reviewing, please take special care to ensure token generation and validation looks correct.

### Preview handling

WordPress doesn't provide a straightforward way to allow previewing a non-public post. It could be possible for us to create a new endpoint for secure previews (e.g. `/vip-workflow-preview/<post-id>/<token>`), but then we'd need to ensure the theme, header, and whole page is rendered in the same way as a preview to a logged-in users. In order to avoid rendering a post page from scratch, I used a solution that keeps almost all of the same WordPress flow used for normal previews. The only difference is a `vw-token` parameter used in the URL.

Previews are made possible [via `allow_secure_preview_requests()`](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/secure-preview.php#L86), which hooks `posts_results`.

The [`post_results` filter](https://github.com/WordPress/WordPress/blob/741a7869fd77a0f979a2116ae87b1413a0e62fb1/wp-includes/class-wp-query.php#L3401) is a part of `WP_Query`'s `get_posts()`, fired after a post preview has been queried and loaded in memory, but before user permissions have been checked. The main logic for checking user permissions [are in these lines of `get_posts()`](https://github.com/WordPress/WordPress/blob/741a7869fd77a0f979a2116ae87b1413a0e62fb1/wp-includes/class-wp-query.php#L3455-L3478), with these being the most important ones:

```php
if ( $post_status_obj && ! $post_status_obj->public ) {
	if ( ! is_user_logged_in() ) {
		// User must be logged in to view unpublished posts.
		$this->posts = array();
	} else { /* ... */ }
```

In short, if a post status is non-public and a user isn't logged in, all queried posts are dropped.

To get around this while using a token, we [check that we're in a preview](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/secure-preview.php#L95) then [verify the `vw-token` parameter](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/secure-preview.php#L101) passed to the page is a valid token. If that passes, we mark the post status as `"publish"` for the current query. After the filter completes, the [WordPress check for non-public statuses](https://github.com/WordPress/WordPress/blob/741a7869fd77a0f979a2116ae87b1413a0e62fb1/wp-includes/class-wp-query.php#L3455) is then skipped, and the preview is allowed to load.

### "Copy Link" button

![secure-preview-modal](https://github.com/user-attachments/assets/6e21d246-8a85-44b2-a852-e8a7978647d4)

The "Copy Link" button [used in the secure preview modal](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/lib/secure-preview.js#L153-L159) is implemented a little strangely. I've added a lot of comments to the [`CopyFromAsyncButton` class](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/lib/components/copy-from-async-button.js) to explain it, but I'll also add some explanation here.

Gutenberg's main way of exposing the clipboard to plugin developers is the [`useCopyToClipboard()` hook](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-compose/#usecopytoclipboard). The first parameter it takes is a static string or function that can generate a string (`() => string`). This hook generates a ref that can be used with a clickable element, and will copy the given text to the clipboard when that element is clicked.

In our case, when the user clicks "Copy Link", we want to [run a fetch in the background](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/lib/secure-preview.js#L98-L102) based on the link parameters the user selected, get the returned URL, and then copy it to the clipboard. However, this doesn't work with `useCopyToClipboard()`, as promises/async functions [are not handled](https://github.com/WordPress/gutenberg/blob/5dae25b15e82a59d28c6c065b5a3745f3f83980f/packages/compose/src/hooks/use-copy-to-clipboard/index.js#L46-L48) as an option, so we have no way to use an async function and the clipboard together.

The result of trying out some different options lead to the creation of [`CopyFromAsyncButton`](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/lib/components/copy-from-async-button.js), which essentially wraps two buttons. When a user clicks the "Copy Link" button:

1. The main button runs `handleCopyClick()`, which `await`s the async function to run.
2. On completion, the result of the async function is set to the `textToCopy` state.
3. When the component rerenders, [`useLayoutEffect()` is used](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/lib/components/copy-from-async-button.js#L29C2-L34) to click a second, invisible button.
4. This invisible button, [which uses the `useCopyToClipboard() hook`](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/lib/components/copy-from-async-button.js#L24-L27), now has `textToCopy` ready to use synchronously.
5. Finally, the async result is copied to the clipboard and the value is passed back to the parent.

If anyone has a better suggestion on this, please let me know! A good core contribution may be to extend `useCopyToClipboard()` to handle async functions so this wouldn't be necessary.

## Testing

See the GIF at the top of this issue for visual reference on where to find the "Secure Preview" button on the sidebar. There are a lot of scenarios to test, and I'll encourage not following an exact set of steps by not providing them. In general for testing, I have two browsers open, one that's logged-in to create posts, and another not logged-in session to verify preview links. Here are some general features to test:

- Test one-time links only work once.

- Test expiration times work as expected. To do this, you can directly edit [the array in `get_link_expiration_options()`](https://github.com/Automattic/vip-workflow-plugin/blob/7f45ac760a0a30325ca2c4f00351c00881ab99a3/modules/secure-preview/secure-preview.php#L132-L147) or use the `vw_secure_preview_expiration_options` filter to add a shorter timeframe, like a one-minute token. Ensure that the token stops working after one minute.
- Try generating multiple tokens at a time, and ensure they work separately.
- Ensure that the "Secure Preview" button only appears for non-public posts in a custom status. The button should also only appear on post types selected under Admin -> VIP Workflow -> Settings:

    ![Screenshot 2024-08-15 at 1 21 58 PM](https://github.com/user-attachments/assets/b07c9d98-c4dd-4bd4-bc78-fc35f6ff03f6)

- Ensure preview links can only be generated for posts that have been saved.
- Test that errors during token generation in the REST endpoint display correctly in the editor. You can edit `modules/secure-preview/rest/secure-preview.php` to return a `WP_Error` during generation, and should see that error exposed in the editor.

alecgeatches committed:
Add sidebar button for preview generation

alecgeatches committed:
Add skeleton secure preview REST endpoint

alecgeatches committed:
Add entry JS and scss build for secure preview module

alecgeatches committed:
Add base secure preview module

alecgeatches committed:
Fix spacing in util.php

alecgeatches committed:
Fix a reference to "edit-flow" in style handle

alecgeatches committed:
Split "secure preview" component into separate Gutenberg plugin registration

alecgeatches committed:
Hit REST endpoint when clicking button, add error notice, move CSS from custom status module

alecgeatches committed:
Add token code in new class, return valid token via REST

alecgeatches committed:
Use 'posts_results' filter with vw-token parameter to get preview URL

alecgeatches committed:
Remove 'preview_id' from REST-generated url as it's unnecessary

alecgeatches committed:
Add @wordpress/block-editor dev dependency

alecgeatches committed:
Transition from secure preview button to URL dropdown on request

alecgeatches committed:
Default token `$expire_on_use` to false

alecgeatches committed:
Add loading spinner on secure preview link

alecgeatches committed:
Add @wordpress/notices to dev packages

alecgeatches committed:
Add easy copy-paste button on generated URL

alecgeatches committed:
Only show secure link generation button when post is in a supported custom status

alecgeatches committed:
Improve error messages when generating secure tokens

alecgeatches committed:
Minor debug, text, and comment changes for PR preparation

alecgeatches committed:
Add modal-based UI for selecting link properties. Still need REST backend changes

alecgeatches committed:
Stop VSCode from deleting unreachable code automatically

alecgeatches committed:
Update REST endpoint and token generation to include custom expirations and one_time_use parameter

alecgeatches committed:
Add @wordpress/editor to dev dependencies

alecgeatches committed:
Add no-cache headers to preview result

alecgeatches committed:
Refactor CopyFromAsyncButton into separate file, add a lot of comments

alecgeatches committed:
Refactor REST expiration logic into utility function

alecgeatches committed:
Prevent expired tokens from appearing to work for logged-in users

alecgeatches committed:
Simplify URL when redirecting permissioned users

alecgeatches committed:
Add REST error for generating secure links before post is saved

alecgeatches committed:
Hide "Generate Link" button on draft posts

alecgeatches committed:
Fix inconsistent ordering of post sections in sidebar

alecgeatches committed:
Minor visual changes to make the preview blend into the sidebar

alecgeatches committed:
Fix error loading editor when the current post type is not selected in settings

alecgeatches commented on jsconfig.json:
```diff
@@ -4,6 +4,7 @@
 		"module": "ES6",
 		"jsx": "preserve",
 		"checkJs": true,
+		"allowUnreachableCode": true
```
Added this because VSCode kept removing code during testing, like if I used an early `return` or commented something out. This doesn't stop ESLint from recognizing the issue, it just stops VSCode from silently obliterating code during a save.

alecgeatches commented on modules/custom-status/lib/custom-status-block.js:
```diff
@@ -3,7 +3,7 @@ import './editor.scss';
 import { compose } from '@wordpress/compose';
 import { dispatch, select, subscribe, withDispatch, withSelect } from '@wordpress/data';
-import { PluginPostStatusInfo } from '@wordpress/edit-post';
+import { PluginPostStatusInfo } from '@wordpress/editor';
```
`PluginPostStatusInfo` from `@wordpress/edit-post` was deprecated in WordPress 6.6 and moved to `@wordpress/editor`. This doesn't appear to have any effect on the component.

ingeniumed replied:
Would this still be backwards compatible though, since we support all the way back to WordPress 6.0 officially?

alecgeatches replied:
@ingeniumed Great point! It looks like this change [was added in WordPress 6.6](https://make.wordpress.org/core/2024/06/18/editor-unified-extensibility-apis-in-6-6/) (from [this PR](https://github.com/WordPress/gutenberg/pull/60814)). I'm thinking we should just bump the WordPress supported version to 6.6, as it'll be what VIP customers are largely using, and we may need to use other 6.1+ features in this plugin for workflow management.

alecgeatches replied:
Did that in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/5116fefd07db581bf8585adf6ac64d953e3ab26e. Let me know if you think that's a bad idea!

ingeniumed replied:
I'd lean towards 6.1 or 6.2 instead, not 6.6 imo

alecgeatches replied:
After internal discussion, I opted to change the minimum down to `6.2` and revert this package name change in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/2d897915463846a3bc83c32386b358c06218009d. Thanks!

alecgeatches committed:
Fix PHPDoc in REST endpoint

chriszarate commented on modules/secure-preview/token.php:
```diff
@@ -0,0 +1,122 @@
+		 *
+		 * @param int    $default_max_lifetime_seconds The default maximum token lifetime in seconds.
+		 */
+		$max_lifetime_seconds = apply_filters( 'vw_secure_preview_max_token_lifetime', $default_max_lifetime_seconds );
```
I'm not sure I really see the point of a maximum, filterable or not, if all of the possible options are filterable via `vw_secure_preview_expiration_options`. It feels like this entire method is superfluous

alecgeatches replied:
I agree! Removed in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/a624afa583a113fdedfd0804ac0c64f1edee9b1c.

chriszarate commented on modules/secure-preview/token.php:
```diff
@@ -0,0 +1,122 @@
+	 */
+	public static function validate_token( $token, $post_id ) {
+		$meta_key     = self::META_KEY;
+		$saved_tokens = get_post_meta( $post_id, $meta_key );
```
`false` is the default, but a good signal that we expect multiple
```suggestion
		$saved_tokens = get_post_meta( $post_id, $meta_key, false );
```

alecgeatches replied:
Addressed in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/4a920c7561c0ac05605fd7cf19c22ea9ad72227d. Curiously I also had to add a type check because explicitly adding `false` triggered [a sniff error](https://github.com/Automattic/VIP-Coding-Standards/blob/develop/WordPressVIPMinimum/Sniffs/Functions/CheckReturnValueSniff.php):

![Screenshot 2024-08-16 at 11 29 08 AM](https://github.com/user-attachments/assets/8f7ce3d5-e60c-4c7a-8da7-0a72780743a3)

chriszarate commented on modules/secure-preview/secure-preview.php:
```diff
@@ -0,0 +1,150 @@
+		if ( Token::validate_token( $token, $posts[0]->ID ) ) {
+			// Temporarily set post_status to 'publish' to stop WP_Query->get_posts() from clearing out
+			// unpublished posts before render
+			$posts[0]->post_status = 'publish';
```
Where does the correct status get restored? This feels risky, since other code (and other hooks) may run before it is restored and incorrectly assume the post is published.

ingeniumed replied:
Would be curious about this as well.

alecgeatches replied:
This is a good concern! I looked through the query code and found that after [the `posts_results` filter](https://github.com/WordPress/WordPress/blob/741a7869fd77a0f979a2116ae87b1413a0e62fb1/wp-includes/class-wp-query.php#L3401) is called and [the logic that allows previewing](https://github.com/WordPress/WordPress/blob/741a7869fd77a0f979a2116ae87b1413a0e62fb1/wp-includes/class-wp-query.php#L3452-L3484) is done executing, [`the_posts` filter is called](https://github.com/WordPress/WordPress/blob/741a7869fd77a0f979a2116ae87b1413a0e62fb1/wp-includes/class-wp-query.php#L3558). I added some code to revert the post status using `the_posts`:

https://github.com/Automattic/vip-workflow-plugin/blob/8cf9f782e01f6a174614e306b3a4a2a2994bbe02/modules/preview/preview.php#L101-L124

Most of the other logic in `WP_Query::get_posts()` between `post_results` and `the_posts` is skipped during a preview, so there's not much opportunity to get unexpected side-effects from other parts of WordPress or plugins.

chriszarate commented on modules/secure-preview/rest/secure-preview.php:
```diff
@@ -0,0 +1,160 @@
+		$expiration_value = $request->get_param( 'expiration' );
+		$is_one_time_use  = $request->get_param( 'is_one_time_use' );
+
+		if ( ! VIP_Workflow::instance()->custom_status->is_post_using_custom_status( $post_id ) ) {
+			$post_status = get_post_status( $post_id );
+
+			if ( 'publish' === $post_status ) {
+				return new WP_Error( 'vip-workflow-published-post', __( 'Secure links can not be generated for published posts.', 'vip-workflow' ) );
+			} elseif ( 'auto-draft' === $post_status ) {
+				return new WP_Error( 'vip-workflow-not-custom-status', __( 'Posts must be saved before a secure link can be generated.', 'vip-workflow' ) );
+			} else {
+				return new WP_Error( 'vip-workflow-not-custom-status', __( 'Secure links can only be generated for pre-published posts with a custom status.', 'vip-workflow' ) );
+			}
+		}
```
Given this validation when the token is created, shouldn't the post status be an input to the token envelope and be part of the token validation? In other words, if the post status changes (e.g., the post is published), shouldn't the token be invalidated?

ingeniumed replied:
Would echo the same Q

alecgeatches replied:
> Given this validation when the token is created, shouldn't the post status be an input to the token envelope and be part of the token validation?

Since we have multiple pre-publish statuses, I'm not sure we want to tie a token to a particular custom status. Let's say we're working with a set of custom statuses like this:

1. Pitch
2. Draft
3. Pending review

It would seem valid to me to create a token for "Pitch" and give it to a colleague, and then maybe promote the post to "Draft" in the meantime. Or the same between "Draft" and "Pending review".

I think the more important part is that a token made for a post in a custom status should only be valid while that post has a custom status assigned to it. What do you think about changing `validate_token()` to ensure that the token is only valid for a post using a custom status, and any other statuses are invalid? This gives some flexibility to modify tracking statuses in a workflow without invalidating links, but ensures they only apply to our custom post statuses.

ingeniumed replied:
i get what you mean, they are intending for one link to be used throughout the process. In the event that they don't want that, then a one time use token could work or the max time could be shortened.

That said, I'm wondering if this makes sense as a setting that customers could toggle, in case they are conscious of security and want a new token. That way, by default it won't get invalidated but if they want to make it stricter they can. This would be similar to the [cascading/restrictive hierarchy filter we have in the governance plugin](https://github.com/Automattic/vip-governance-plugin?tab=readme-ov-file#vip_governance__is_block_allowed_in_hierarchy).

alecgeatches replied:
I added some code to explicitly ignore and clear tokens when a preview for a non-custom-status or non-supported post type is used:

https://github.com/Automattic/vip-workflow-plugin/blob/84e5bc42e46db0926a32652a206a47f8d5d0a37f/modules/preview/token.php#L57-L64

Here's how that will work with some different situations:

1. If a preview URL is generated for a post with custom status "Pitch", and then it moves to another custom status "Draft", the key will continue to work.
2. If a post has a token URL generated, and then moves to a non-custom status (either "Publish" or another custom status that isn't managed by our plugin), the key will be rejected and all tokens will be removed.
3. If a post has a token URL generated in a custom status, then briefly moves to "Publish", and then moves back to a custom statues (e.g. "Pending Review"), the URL will continue to work, as long as it wasn't used while the post was in the "Publish" state.

The third situation above may seem a little odd, but I think it's not harmful if this situation occurs. I opted to do key removal in the `validate_token()` function rather than relying on an external `transition_post_status` or similar filter to remove tokens automatically in the background. The main reasoning is that this simplifies the logic - we only need to think about token management in one function, and there are no other side-effects that will also manage tokens. Also, just in case a post doesn't fire the expected hooks (like if a post was updated to `publish` directly in the database), we'll still need this logic in `validate_token()` anyhow.

This solution keeps "security"-related logic in a single flow in one function, which I think is worthwhile.

chriszarate reviewed (commented):
I left some comments around the token generation and validation. However, I worry about labeling this a "secure preview." 

Visiting a "secure" link generally implies that (WP) user authentication and authorization have taken place, and these bearer tokens are subject to none. Calling them secure may encourage people to select long lifetimes and may contribute to misunderstandings about how they work. 

Can we find a better way to communicate that these bearer tokens allow _anyone_ to view the unpublished post? For example, Google often describes this access model as "Anyone with the link," which neatly describes that the URL itself allows access.

Might we also display a warning with "expire on use" is disabled, especially with a longer token lifetime?

smithjw1 commented:
> Can we find a better way to communicate that these bearer tokens allow _anyone_ to view the unpublished post? For example, Google often describes this access model as "Anyone with the link," which neatly describes that the URL itself allows access.

I think I've been calling this a "secure preview" because that's what customers often call it. I completely agree that it isn't "secure." I think we just drop the word secure and call it a "Preview Link."

> Might we also display a warning with "expire on use" is disabled, especially with a longer token lifetime?

I think some micro copy on the dialog is a good idea:

"Anyone with this link will be able to preview the post."

ingeniumed commented on modules/custom-status/custom-status.php:
```diff
@@ -1239,6 +1239,27 @@ public function workflow_is_publish_blocked( $post_id ) {
+			return false;
+		}
+
+		$custom_post_types = $this->get_post_types_for_module();
```
Minor: We should rename this function because it's not module specific anymore.

alecgeatches replied:
True! Changed to `get_supported_post_types()` in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/f792715e4aa735a2a8119e77fe429f1b5931854e.

ingeniumed commented on modules/secure-preview/lib/secure-preview.js:
```diff
@@ -0,0 +1,240 @@
+	const { getEditedPostAttribute, getCurrentPostType, getCurrentPost } = select( editorStore );
+
+	const post = getCurrentPost();
+	const isDraft = post?.status === 'auto-draft';
```
I'd a comment here explaining why this is done this way, and change the name of this var to be `isAutoDraft` instead.

alecgeatches replied:
Agree that `isDraft` is a confusing name, especially given that we have a "Draft" status. I changed it to `isUnsavedPost` in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/ac739e071f6a077d3c63728c42cf1b639149d23c which makes the usage clearer in the plugin, and also added a comment.

ingeniumed commented on modules/secure-preview/rest/secure-preview.php:
```diff
@@ -0,0 +1,160 @@
+<?php
+/**
+ * class EditStatus
```
Should change this name and the description below

alecgeatches replied:
Changed in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/283371c204bacd6e1cf3f6fc2c4d301ed4416f56.

ingeniumed commented on modules/secure-preview/secure-preview.php:
```diff
@@ -0,0 +1,150 @@
+/**
+ * The secure preview module, for sharing pre-published content.
+ *
+ * @package vip-bundle-decoupled
```
Would remove this

ingeniumed commented on modules/secure-preview/secure-preview.php:
```diff
@@ -0,0 +1,150 @@
+<?php
+/**
+ * The secure preview module, for sharing pre-published content.
```
Minor spacing issue before the for

ingeniumed commented on modules/secure-preview/token.php:
```diff
@@ -0,0 +1,122 @@
+			}
+
+			// Check if token matches. If it does, mark as expired.
+			if ( ! $is_expired && $token === $saved['token'] && self::TOKEN_ACTION === $saved['action'] ) {
```
Would be worth expiring the token this way on the status change action that we hook into, in the custom status module.

ingeniumed commented on modules/secure-preview/token.php:
```diff
@@ -0,0 +1,122 @@
+	 * @param  int    $post_id Post ID.
+	 * @return bool
+	 */
+	public static function validate_token( $token, $post_id ) {
```
Would this token show up in the rest api for a post? Guessing no 🤔

alecgeatches replied:
Interesting question! After a bit of research, it appears it will not be available in REST unless we explicitly register the token meta key via [`register_meta()`](https://developer.wordpress.org/reference/functions/register_meta/) and set `'show_in_rest' => true` as an argument. The post type itself must also[ register support for `custom-fields`](https://developer.wordpress.org/reference/functions/add_post_type_support/), so this shouldn't be exposed unless someone jumps through some hoops to do so.

ingeniumed reviewed (changes requested):
- Ran into a weird bug where when the post is first made the timeout value is 1 hour, but if I change the status using the extended status dropdown it changes to 8 hours.
- Should update the README as part of this, to add in a brief explanation of this feature and perhaps a gif as well. This way when we are ready for a beta we don't need to go in and update missing documentation.

alecgeatches committed:
Bump minimum supported WordPress version to 6.6

alecgeatches committed:
Hard-code generate_token() to require the "edit_posts" capability

alecgeatches committed:
Explicitly pass single = false in get_ post_meta() for clarity in token logic

alecgeatches committed:
Replace "Secure Preview" with "Preview" in code and UI

alecgeatches commented on modules/secure-preview/token.php:
```diff
@@ -0,0 +1,122 @@
+	 * @return array|WP_Error
+	 */
+	public static function generate_token( $post_id, $cap, $is_one_time_use, $expiration_seconds ) {
+		if ( ! current_user_can( $cap, $post_id ) ) {
```
That's a good idea for security. Fixed in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/a624afa583a113fdedfd0804ac0c64f1edee9b1c.

alecgeatches committed:
Change `get_post_types_for_module()` to `get_supported_post_types()`, as these are shared between modules now.

alecgeatches committed:
Rename `isDraft` to `isUnsavedPost` to be less confusing and add a comment

alecgeatches committed:
Update copy-pasted class header

alecgeatches committed:
Move required WordPress version back to 6.2, import PluginPostStatusInfo from @wordpress/edit-post

alecgeatches committed:
Fix duplicated WP_Error code

alecgeatches committed:
Fix incorrect import update in `2d89791`

alecgeatches committed:
Reset post status after preview logic using 'the_posts' filter

alecgeatches committed:
Update the_posts filter undo function to run at a higher priority than default

alecgeatches committed:
Add warning to preview dialog that anyone with a preview link can preview a post

alecgeatches committed:
Add logic to explicitly clear preview tokens on non-custom status posts

alecgeatches committed:
Redirect to public post URLs

alecgeatches committed:
Add README sections about preview link generation and filter

alecgeatches committed:
Fix markdown image links

alecgeatches commented:
@chriszarate @ingeniumed @smithjw1 Thank you all for your excellent feedback. This is ready for another review! I've responded to individual comments above, but here were the main changes:

- Mentions of a "Secure Link" have been replaced by "Preview Link". I've also added the suggested copy to the link dialog:

    ![Screenshot 2024-08-19 at 12 47 02 PM](https://github.com/user-attachments/assets/92c85575-fb12-4338-a261-836418660f54)

- We now restore a preview post's status after temporarily setting it to `publish`. [See the discussion here](https://github.com/Automattic/vip-workflow-plugin/pull/19#discussion_r1722190181).
- We now explicitly delete tokens when previewing a post outside of the plugin's supported custom statuses. [See the discussion here](https://github.com/Automattic/vip-workflow-plugin/pull/19#discussion_r1722242053).
- Added a "Preview Links" and `vw_preview_expiration_options` filter documentation [to the README](https://github.com/Automattic/vip-workflow-plugin/blob/add/secure-preview-links/README.md).

ingeniumed commented on modules/preview/lib/preview.js:
```diff
@@ -0,0 +1,250 @@
+/**
+ * Custom component to generate and copy a preview link in the post sidebar.
+ */
+const VIPWorkflowPreview = ( { status, postType, isUnsavedPost } ) => {
```
Could you add some more detailed comments in this plugin? It'll make it easier to revisit this, down the line?

alecgeatches replied:
I did an attempt at this in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/77c252587063e3ca1f48d1ddcd498db94ae080a1, but didn't have a lot of comments to add. Most of this code is pretty idiomatic for functional React components, and I'm not sure where most comments will provide context other than re-stating the code. I've already attempted to isolate non-idiomatic code like [`<CopyFromAsyncButton>`](https://github.com/Automattic/vip-workflow-plugin/blob/77c252587063e3ca1f48d1ddcd498db94ae080a1/modules/preview/lib/components/copy-from-async-button.js) in a separate well-documented component.

If you have any specific areas that you feel are lacking, please let me know!

alecgeatches replied:
Added a couple more comments in https://github.com/Automattic/vip-workflow-plugin/pull/19/commits/ff5481791181c42aa4f7a85df8bea2130e5cd746.

ingeniumed reviewed (approved):
Minor comments, otherwise it LGTM. Nice work on this 👍🏾

alecgeatches committed:
Add ability to generate another preview link without refreshing page

alecgeatches committed:
Add some comments, tiny refactoring of preview plugin component

alecgeatches committed:
Add component headers
